File: PMystic/mystic/environment.py
from runtime_error import RTE


class Environment:
    def __init__(self, enclosing=None):
        self.__values = {}
        self.__enclosing = enclosing

    def define(self, name, value):
        self.__values[name] = value

    def get(self, name):
        if name.lexeme in self.__values:
            return self.__values[name.lexeme]

        if self.__enclosing is not None:
            return self.__enclosing.get(name)

        raise RTE(name, f"Undefined variable '{name.lexeme}'.")

    def get_at(self, dist, name):
        return self.__ancestor(dist).__values[name.lexeme]

    def assign(self, name, value):
        if name.lexeme in self.__values:
            self.__values[name.lexeme] = value
            return

        if self.__enclosing is not None:
            self.__enclosing.assign(name, value)
            return

        raise RTE(name, f"Undefined variable '{name.lexeme}'.")

    def assign_at(self, dist, name, value):
        self.__ancestor(dist).__values[name.lexeme] = value

    def __ancestor(self, dist):
        env = self
        for i in range(dist):
            env = env.__enclosing
        return env

File: JMystic/mystic/Environment.java
package JMystic.mystic;

import java.util.HashMap;
import java.util.Map;

/**
 * Environment
 */
public class Environment {
    final Environment enclosing;
    private final Map<String, Object> values = new HashMap<>();

    Environment() {
        enclosing = null;
    }

    Environment(Environment enclosing) {
        this.enclosing = enclosing;
    }

    void define(String name, Object value) {
        values.put(name, value);
    }

    Object get(Token name) {
        if (values.containsKey(name.lexeme))
            return values.get(name.lexeme);

        if (enclosing != null)
            return enclosing.get(name);

        throw new RuntimeError(name, "Undefined variable '" + name.lexeme + "'.");
    }

    Object getAt(int dist, String name) {
        return ancestor(dist).values.get(name);
    }

    void assign(Token name, Object value) {
        if (values.containsKey(name.lexeme)) {
            values.put(name.lexeme, value);
            return;
        }

        if (enclosing != null) {
            enclosing.assign(name, value);
            return;
        }

        throw new RuntimeError(name, "Undefined variable '" + name.lexeme + "'.");
    }

    void assignAt(int dist, Token name, Object value) {
        ancestor(dist).values.put(name.lexeme, value);
    }

    Environment ancestor(int dist) {
        Environment env = this;
        for (int i = 0; i < dist; i++)
            env = env.enclosing;

        return env;
    }
}